google-beta/resource_pubsub_subscription.go

```diff
@@ -16,42 +34,38 @@ func resourcePubsubSubscription() *schema.Resource {
 		Delete: resourcePubsubSubscriptionDelete,
 
 		Importer: &schema.ResourceImporter{
-			State: resourcePubsubSubscriptionStateImporter,
+			State: resourcePubsubSubscriptionImport,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(240 * time.Second),
+			Update: schema.DefaultTimeout(240 * time.Second),
+			Delete: schema.DefaultTimeout(240 * time.Second),
 		},
 
 		Schema: map[string]*schema.Schema{
 			"name": {
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: comparePubsubSubscriptionBasename,
 			},
-
 			"topic": {
 				Type:             schema.TypeString,
 				Required:         true,
 				ForceNew:         true,
-				DiffSuppressFunc: compareSelfLinkOrResourceName,
+				DiffSuppressFunc: comparePubsubTopicBasename,
```

Comment:
Since this does effectively the same thing as `compareSelfLinkOrResourceName`, I'm inclined to keep using `compareSelfLinkOrResourceName` and not introduce new resource / product-specific validation.

Reply:
I originally considered using `compareSelfLinkOrResourceName` but that does not work since it assumes [the internal state is the long form](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/third_party/terraform/utils/self_link_helpers.go#L33) when we actually store the short form due to the `replaceVars`. I then considered using `getRelativePath` assuming it would give me either the last part of a long form or a short form by itself, but it will only work fine with the first. For generalizing, we need a compare version that only considers the equivalent of what the UNIX `basename(1)` would return, but we don't seem to have that, do we?

Reply:
Correction, this comment applies to the `name` attribute below, for `topic` it may be fine as we store the long version.

Reply:
Have you found it to be the case that the diff suppress didn't work when `old` was just a name? I don't think it's actually true that it requires the value stored in state to be the full link;

* Terraform will compare the two strings, so if they're both the same we don't show a diff regardless
* If `new` is just a name, we compare `new` to the name of `old` https://github.com/GoogleCloudPlatform/magic-modules/blob/master/third_party/terraform/utils/self_link_helpers.go#L40
* If _both_ `new` and `old` are full/relative paths, we compare by relative path with https://github.com/GoogleCloudPlatform/magic-modules/blob/master/third_party/terraform/utils/self_link_helpers.go#L15

Reply:
Yes, I've tested it. Before we had the full form in the state so that was not an issue. Now it's the short form due to `replaceVars`:
```
$ git diff products/pubsub/terraform.yaml
diff --git a/products/pubsub/terraform.yaml b/products/pubsub/terraform.yaml
index 3b5f48a9..35e232de 100644
--- a/products/pubsub/terraform.yaml
+++ b/products/pubsub/terraform.yaml
@@ -24,7 +24,7 @@ overrides: !ruby/object:Overrides::ResourceOverrides
           topic_name: "example-topic"
     properties:
       name: !ruby/object:Overrides::Terraform::PropertyOverride
-        diff_suppress_func: 'comparePubsubTopicBasename'
+        diff_suppress_func: 'compareSelfLinkOrResourceName'
         custom_expand: templates/terraform/custom_expand/computed_topic_name.erb
         custom_flatten: templates/terraform/custom_flatten/name_from_self_link.erb
   Subscription: !ruby/object:Overrides::Terraform::ResourceOverride
```
with the proposed change above, the fullName test fails:
```
$ make testacc TESTARGS='-run=TestAccPubsubTopic_fullName'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccPubsubTopic_fullName -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google-beta/version.ProviderVersion=acc"
?   	github.com/terraform-providers/terraform-provider-google-beta	[no test files]
=== RUN   TestAccPubsubTopic_fullName
=== PAUSE TestAccPubsubTopic_fullName
=== CONT  TestAccPubsubTopic_fullName
--- FAIL: TestAccPubsubTopic_fullName (4.09s)
    testing.go:538: Step 0 error: After applying this step, the plan was not empty:
        
        DIFF:
        
        DESTROY/CREATE: google_pubsub_topic.foo
          name:    "tf-test-topic-4et3q9qkgn" => "projects/drebes-playground-alpha-5ae4/topics/tf-test-topic-4et3q9qkgn" (forces new resource)
          project: "drebes-playground-alpha-5ae4" => "<computed>"
        
        STATE:
        
        google_pubsub_topic.foo:
          ID = projects/drebes-playground-alpha-5ae4/topics/tf-test-topic-4et3q9qkgn
          provider = provider.google
          labels.% = 0
          name = tf-test-topic-4et3q9qkgn
          project = drebes-playground-alpha-5ae4
FAIL
FAIL	github.com/terraform-providers/terraform-provider-google-beta/google-beta	4.125s
?   	github.com/terraform-providers/terraform-provider-google-beta/scripts/affectedtests	[no test files]
?   	github.com/terraform-providers/terraform-provider-google-beta/version	[no test files]
make: *** [testacc] Error 1
```

Reply:
Oh! I'd confused myself and thought we were commenting on `name`, my apologies. We stored the full URI for `topic` and should continue to do that, sorry about the confusion on my part.